gds/src/main/scala/edu/gemini/aspen/gds/observations/ObservationManager.scala

```diff
@@ -0,0 +1,112 @@
+import java.util.logging.Logger
+import scala.concurrent.duration._
+
+trait ObservationManager[F[_]] {
+  def process(stateEvent: ObservationStateEvent): F[Unit]
+}
+
+object ObservationManager {
+  private val logger = Logger.getLogger(this.getClass.getName)
+
+  // TODO: Make this a configuration value.
+  val lifetime = 20.minutes
+
+  def apply[F[_]: Async](
+    keywordManager: KeywordManager[F],
+    obsStateQ:      QueueSink[F, ObservationStateEvent]
+  ): F[ObservationManager[F]] =
+    MapRef.ofConcurrentHashMap[F, DataLabel, ObservationItem[F]]().map { mapref =>
+      new ObservationManager[F] {
+        def process(stateEvent: ObservationStateEvent): F[Unit] = stateEvent match {
+          case Start(dataLabel, programId) =>
+            for {
+              _   <- logger.infoF(s"Starting observation $dataLabel")
+              now <- Clock[F].realTime
+              fsm <-
+                ObservationFSM(dataLabel, obsStateQ)
+              // check to be sure it doesn't already exist? Or, should we just start over, anyway. Maybe a warning, at least.
+              _   <- mapref.setKeyValue(dataLabel, ObservationItem(programId, fsm, now + lifetime))
+              _   <- keywordManager.initialize(dataLabel)
+            } yield ()
+
+          case e @ Stop(dataLabel) => withObsItem(dataLabel, e)(_.fsm.stopObservation)
+
+          case Abort(dataLabel) =>
+            logger.warningF(s"Received Abort event for observation $dataLabel") >>
+              obsStateQ.offer(Delete(dataLabel))
+
+          case Delete(dataLabel) =>
+            logger.infoF(s"Removing observation $dataLabel") >> keywordManager.delete(
+              dataLabel
+            ) >> mapref.unsetKey(dataLabel)
+
+          case e @ AddObservationEvent(dataLabel, obsEvent) =>
+            withObsItem(dataLabel, e)(_.fsm.addObservationEvent(obsEvent))
+
+          case e @ AddKeyword(dataLabel, keyword) =>
+            withObsItem(dataLabel, e)(_ => keywordManager.add(dataLabel, keyword))
+
+          case e @ CollectKeywords(dataLabel, obsEvent) =>
+            withObsItem(dataLabel, e)(_ => keywordManager.collect(dataLabel, obsEvent))
+
+          case e @ Step(dataLabel) => withObsItem(dataLabel, e)(_.fsm.step)
+
+          case e @ Complete(dataLabel) =>
+            withObsItem(dataLabel, e) { _ =>
+              for {
+                kws <- keywordManager.get(dataLabel)
+                _   <- logger.infoF(s"Got these keywords for $dataLabel:\n\t${kws.mkString("\n\t")}")
+                _   <- logger.infoF(
+                         "Here is where we would deal with the keywords and transfer the fits file."
+                       )
+                _   <- obsStateQ.offer(Delete(dataLabel))
+              } yield ()
+            }
+
+          case PurgeStale =>
+            for {
+              _    <- logger.infoF("Checking for expired observations.")
+              keys <- mapref.keys
+              now  <- Clock[F].realTime
```

Comment:
Does `realTime` give you a `FiniteDuration`? That seems odd f something called `realTime`

Reply:
Yeah. That surprised me as well. It is a `FiniteDuration` with a duration equivalent to `System.currentTimeMillis`.